compiler/ilgen/OMRIlBuilder.cpp

```diff
@@ -2545,26 +2545,108 @@ OMR::IlBuilder::Switch(const char *selectionVar,
    TR::Node *defaultNode = TR::Node::createCase(0, (*defaultBuilder)->getEntry()->getEntry());
    TR::Node *lookupNode = TR::Node::create(TR::lookup, numCases + 2, loadValue(selectorValue), defaultNode);
 
+   generateSwitchCases(lookupNode, defaultNode, defaultBuilder, numCases, cases);
+   }
+
+void
+OMR::IlBuilder::Switch(const char *selectionVar,
+                  TR::IlBuilder **defaultBuilder,
+                  uint32_t numCases,
+                  ...)
+   {
+   va_list args;
+   va_start(args, numCases);
+   JBCase **cases = createCaseArray(numCases, args);
+   va_end(args);
+
+   Switch(selectionVar, defaultBuilder, numCases, cases);
+   }
+
+void
+OMR::IlBuilder::TableSwitch(const char *selectionVar,
+                  TR::IlBuilder **defaultBuilder,
+                  bool generateBoundsCheck,
+                  uint32_t numCases,
+                  JBCase **cases)
+   {
+   TR::IlValue *selectorValue = Load(selectionVar);
+   TR_ASSERT(selectorValue->getDataType() == TR::Int32, "TableSwitch only supports selector having type Int32");
+   if (numCases > 0)
+      {
+      int32_t low = cases[0]->_value;
+      if (low != 0)
+         selectorValue = Sub(selectorValue, ConstInt32(low));
+      }
+
+   *defaultBuilder = createBuilderIfNeeded(*defaultBuilder);
+
+   TR::Node *defaultNode = TR::Node::createCase(0, (*defaultBuilder)->getEntry()->getEntry());
+   TR::Node *tableNode = TR::Node::create(TR::table, numCases + 2, loadValue(selectorValue), defaultNode);
+   if (!generateBoundsCheck)
+       tableNode->setIsSafeToSkipTableBoundCheck(true);
+
```

Comment:
I missed this one in the first review too :( .

what happens if the cases provided aren't dense? i.e. if there's a missing label ?

I believe the table opcode assumes the cases are dense, so depending on how friendly you want to be, we could either 1) assert if a case value is missing, or 2) automatically create a filler case using the default builder (I lean towards this one, but I can imagine ugly things that could happen, so an argument for #1 could convince me to go that way).

I think we have to do something here... and maybe the TableSwitch sample should be non-dense if you like the second case better :) .

Reply:
I added the assert. Adding the cases feels wrong to me and it would mean that the helper could not be shared between Switch and TableSwitch.

Reply:
you'd only need to allocate a new array of `JBCase *`s and fill them in; you could still share the helper.

I think what you have is a reasonable starting point; if people start using it and find the other approach more valuable then we can easily change the behaviour or introduce yet another service (`SparseTableSwitch` :) )